### PR TITLE
Refactor Font::DetermineLines.

### PR DESCRIPTION
**Pull request by LGM-Doyle**
This PR refactors Font::DetermineLines.
It creates FillTextElements to do the intensive processing of DetermineLines.
It moves all of the boost::xpressive regex creation into CompiledRegex, to reduce repetition in the code.

### CODE REVIEW

File: GG/src/Font.cpp [hunk 1]
```diff
@@ -240,6 +240,21 @@ namespace {
...
+    {
+        typedef void result_type;
+        void operator()(const std::string* str,
+                        std::stack<Font::Substring>& tag_stack,
+                        bool& ignore_tags,
+                        const boost::xpressive::ssub_match& sub) const
+        {
+            tag_stack.push(Font::Substring(*str, sub));
+            if (tag_stack.top() == PRE_TAG)
+                ignore_tags = true;
```
Comment (adrianbroher) [inline GG/src/Font.cpp:257]: `ignore_tags` may break the stack logic in case of nested `<pre>` tags.

File: GG/src/Font.cpp [hunk 2]
```diff
@@ -685,33 +700,167 @@ Font::Glyph::Glyph(const boost::shared_ptr<Texture>& texture, const Pt& ul, cons
...
-        ActionTags().insert("u");
-        ActionTags().insert("rgba");
-        ActionTags().insert(ALIGN_LEFT_TAG);
-        ActionTags().insert(ALIGN_CENTER_TAG);
-        ActionTags().insert(ALIGN_RIGHT_TAG);
-        ActionTags().insert(PRE_TAG);
-
-        // Always know the action tags.
-        KnownTags().insert(ActionTags().begin(), ActionTags().end());
+        StaticTagHandler().Insert("i");
```
Comment (adrianbroher) [inline GG/src/Font.cpp]: This stems from old code but while you're fixing this already: the literals should be replaced with constants like it was already done for ALIGN_LEFT_TAG and relatives.

File: GG/src/Font.cpp [hunk 3]
```diff
@@ -1097,61 +1242,29 @@ void Font::ThrowBadGlyph(const std::string& format_str, boost::uint32_t c)
...
-#endif
-
-    std::vector<boost::shared_ptr<TextElement> > local_text_elements;
-
-    std::vector<boost::shared_ptr<TextElement> >& text_elements =
-        text_elements_ptr ? *text_elements_ptr : local_text_elements;
-
+void Font::FillTextElements(const std::string& text,
+                            bool ignore_tags,
+                            std::vector<boost::shared_ptr<TextElement> >& text_elements) const {
```
Comment (adrianbroher) [inline GG/src/Font.cpp]: Missing line break before brace.

File: GG/src/Font.cpp [hunk 4]
```diff
@@ -1097,61 +1242,29 @@ void Font::ThrowBadGlyph(const std::string& format_str, boost::uint32_t c)
...
     if (text_elements.empty()) {
         using namespace boost::xpressive;
 
-        std::stack<Substring> tag_stack;
-        bool ignore_tags = format & FORMAT_IGNORETAGS;
-        MatchesKnownTag matches_known_tag(KnownTags(), ignore_tags);
-        MatchesTopOfStack matches_tag_stack(tag_stack, ignore_tags);
+        // Fetch and use the regular expression from the TagHandler which parses all the known XML tags.
+        sregex & regex = StaticTagHandler().Regex(text, ignore_tags);
```
Comment (adrianbroher) [inline GG/src/Font.cpp]: Unnecessary space in front of ampersand.

File: GG/src/Font.cpp [hunk 5]
```diff
@@ -1097,61 +1242,29 @@ void Font::ThrowBadGlyph(const std::string& format_str, boost::uint32_t c)
...
+        sregex_iterator it(text.begin(), text.end(), regex);
 
+        // These are the types found by the regular expression: XML open/close tags, text and
```
Comment (adrianbroher) [inline GG/src/Font.cpp:1260]: No need for change but those are not XML tags as they don't follow the specification for XML especially for attributes but 'something else'.

**Review by adrianbroher — CHANGES_REQUESTED** [review]
Looks good. The stack should work properly except for nested `pre` tags.  Can you squash the style changes into the commits that were intoducing the style violations?

**Comment by adrianbroher** [comment]
Thanks a lot for breaking the PR up.  That made it much easier to review it properly.  Could you also split     32b5149 into a 'update comments' commit and squash the style changes into the corresponding commits that introduced the style violation?

File: GG/src/Font.cpp (hunk 1 above)
Comment (LGM-Doyle) [reply GG/src/Font.cpp:257]: That is correct. This is the existing behavior.  It is also documented in the comments of Font.h

File: GG/src/Font.cpp (hunk 2 above)
Comment (LGM-Doyle) [reply GG/src/Font.cpp]: I added the new tags, as well as documentation that "s" for shadow is a recognized tag.

File: GG/src/Font.cpp (hunk 5 above)
Comment (LGM-Doyle) [reply GG/src/Font.cpp:1260]: And this is not an XML parser.  A "correct" XML is supposed to stop processing completely and report an error if anything about the document is incorrect.  

However, throughout freeorion's technical documentation we are referring to the various ad-hoc formats as XML, so I chose not to start referring to them as XML-like here.

**Comment by LGM-Doyle** [comment]
I left the style and comment changes in a commit referencing the reviews.

It recognizes the work put in by the reviewers as valuable to the project and is worth recording in the commit log.

File: GG/src/Font.cpp (hunk 1 above)
Comment (adrianbroher) [reply GG/src/Font.cpp:257]: Hm, time to file a bug then.

File: GG/src/Font.cpp (hunk 2 above)
Comment (adrianbroher) [reply GG/src/Font.cpp]: Thanks. :+1:

**Comment by adrianbroher** [comment]
> I left the style and comment changes in a commit referencing the reviews.
> 
> It recognizes the work put in by the reviewers as valuable to the project and is worth recording in the commit log.

We won't come to an agreement here.  But that's nothing that should prevent merging this PR.  Compiling and giving it a last test run before merging.

**Review by adrianbroher — APPROVED** [review]

**Comment by adrianbroher** [comment]
So, which one is next? #1030? Can you rebase it?

**Comment by LGM-Doyle** [comment]
#1030 thru #1033 are all in parallel and have been re-based.

**Pull request merged**